File: src/Distributed/MortonOrderRedistributeNodes.cc
//---------------------------------Spheral++----------------------------------//
// MortonOrderRedistributeNodes
// Attempt to redistribute nodes such that they are laid out in memory
// in a Morton ordering.  Note that this involves renumbering the nodes of 
// each NodeList, not just redistributing them between processors.
//
// Created by JMO, Tue Mar 25 14:19:18 PDT 2008
//----------------------------------------------------------------------------//
#include "MortonOrderRedistributeNodes.hh"
#include "Utilities/mortonOrderIndices.hh"
#include "DataBase/DataBase.hh"
#include "Field/FieldList.hh"

namespace Spheral {

//------------------------------------------------------------------------------
// Constructor.
//------------------------------------------------------------------------------
template<typename Dimension>
MortonOrderRedistributeNodes<Dimension>::
MortonOrderRedistributeNodes(double dummy,
                             const double minNodesPerDomainFraction,
                             const double maxNodesPerDomainFraction,
                             const bool workBalance,
                             const bool localReorderOnly):
  SpaceFillingCurveRedistributeNodes<Dimension>(dummy, 
                                                minNodesPerDomainFraction,
                                                maxNodesPerDomainFraction,
                                                workBalance,
                                                localReorderOnly) {
}

//------------------------------------------------------------------------------
// Destructor
//------------------------------------------------------------------------------
template<typename Dimension>
MortonOrderRedistributeNodes<Dimension>::
~MortonOrderRedistributeNodes() {
}

//------------------------------------------------------------------------------
// Hash the node positions into their tree ordered indices.
//------------------------------------------------------------------------------
template<typename Dimension>
FieldList<Dimension, KeyTraits::Key>
MortonOrderRedistributeNodes<Dimension>::
computeHashedIndices(const DataBase<Dimension>& dataBase) const {
  return mortonOrderIndices(dataBase);
}

}

File: src/Pybind11Wraps/Distributed/MortonOrderRedistributeNodes.py
#-------------------------------------------------------------------------------
# MortonOrderRedistributeNodes
#-------------------------------------------------------------------------------
from PYB11Generator import *
from SpaceFillingCurveRedistributeNodes import *

@PYB11template("Dimension")
class MortonOrderRedistributeNodes(SpaceFillingCurveRedistributeNodes):
    """MortonOrderRedistributeNodes

Attempt to redistribute nodes such that they are laid out in memory
in a Morton ordering.  Note that this involves renumbering the nodes of 
each NodeList, not just redistributing them between processors.

Warren & Salmon (1995), Computer Physics Communications, 87, 266-290."""

    PYB11typedefs = """
    typedef typename KeyTraits::Key Key;
"""

    #...........................................................................
    # Constructors
    def pyinit(self,
               dummy = "const double",
               minNodesPerDomainFraction = ("const double", "0.5"),
               maxNodesPerDomainFraction = ("const double", "1.5"),
               workBalance = ("const bool", "true"),
               localReorderOnly = ("const bool", "false")):
        "Constructor"

    #...........................................................................
    # Virtual methods
    @PYB11virtual
    @PYB11const
    def computeHashedIndices(self,
                             dataBase = "const DataBase<%(Dimension)s>&"):
        "This is the required method for all descendant classes."
        return "FieldList<%(Dimension)s, Key> "
